requires.py

```diff
@@ -1,113 +1,288 @@
+        """
+        return self.server_certs_map
+
+    @property
+    def client_certs(self):
+        """
+        List of [Certificate][] instances for all available client certs.
+        """
         name = hookenv.local_unit().replace('/', '_')
-        conversation = self.conversation()
-        reqs = conversation.get_remote('{}.processed_requests'.format(name))
-        if reqs:
-            return json.loads(reqs)
+        field = '{}.processed_client_requests'.format(name)
+        certs_data = self.all_joined_units.received[field] or {}
+        return [Certificate('client',
+                            common_name,
+                            cert['cert'],
+                            cert['key'])
+                for common_name, cert in certs_data.items()]
+
+    @property
+    def client_certs_map(self):
+        """
+        Mapping of client [Certificate][] instances by their `common_name`.
+        """
+        return {cert.common_name: cert for cert in self.client_certs}
+
+    def request_server_cert(self, cn, sans=None, cert_name=None):
+        """
+        Request a server certificate and key be generated for the given
+        common name (`cn`) and optional list of alternative names (`sans`).
+
+        The `cert_name` is deprecated and not needed.
+
+        This can be called multiple times to request more than one server
+        certificate, although the common names must be unique.  If called
+        again with the same common name, it will be ignored.
+        """
+        # assume we'll only be connected to one provider
+        to_publish_json = self.relations[0].to_publish
+        to_publish_raw = self.relations[0].to_publish_raw
+        if not to_publish_raw['common_name']:
+            # for backwards compatibility, first request goes in its own fields
+            to_publish_raw['common_name'] = cn
+            to_publish_json['sans'] = sans or []
+            to_publish_raw['certificate_name'] = cert_name
         else:
-            return {}
+            # subsequent requests go in the collection
+            requests = to_publish_json.get('cert_requests', {})
+            requests[cn] = {'sans': sans or []}
+            to_publish_json['cert_requests'] = requests
+
+    def add_request_server_cert(self, cn, sans):
+        """
+        Deprecated.  Use [request_server_cert][] instead.
+        """
+        self.request_server_cert(cn, sans)
+
+    def request_server_certs(self):
+        """
+        Deprecated.  Just use [request_server_cert][]; this does nothing.
+        """
+        pass
+
+    def request_client_cert(self, cn, sans):
+        """
+        Request a client certificate and key be generated for the given
+        common name (`cn`) and list of alternative names (`sans`).
+
+        This can be called multiple times to request more than one client
+        certificate, although the common names must be unique.  If called
+        again with the same common name, it will be ignored.
```

Comment:
No, the new request will be ignored, because of how the provides side detects whether a given request has been handled (i.e., is there a matching cert with the same common name?  if so, it's not new).  I also added the TODO to have it detect changes to the SANs and issue a new cert.

Reply:
I think this might be an issue when considering the [`extra_sans`](https://github.com/juju-solutions/bundle-canonical-kubernetes/wiki/Certificate-regeneration-via-extra_sans-options) option, but I'm not certain that the refactor actually changed this behavior.  If this is going to be a concern, feel free to NACK this and I will work on the TODO.